### PR TITLE
Identifier support using variablesReference

### PR DESCRIPTION
**Pull request by lochrist**
- Add sendDebuggerRequest
- Engine request can contain requestId
- Refactor all lua helpers in the same block
- Create lua helper to report the result of an evaluation to vscode
- Keep scopes and variablesReference in memory. They are clear when we received a new callstack.
- Ensure all places evaluating variables have the same code path: check for an identifier than can resolve in our scopes and populate the pretty expandable ui. If the expression cannot be evaluated to an identifier: print the result of the evaluation.
- repl evaluation will pretty print identifiers in the console
- hover will pretty print identifiers in the console
- watch: will potentially pretty print identifier. But since they are implemented BEFORE we get the stack trace they might use normal evaluation.
- up_values are display (in a Closure block)
- modify_variable works

### CODE REVIEW

**Comment by lochrist** [comment]
@jschmidt42 please review. I can setup a live demo tomorrow.

File: src/stingray-debugger.ts [hunk 1]
```diff
@@ -440,25 +672,162 @@ class StingrayDebugSession extends DebugSession {
...
+
+        // Evaluate lua script on the engine side.
+        let evalScript = `evaluate_script_expression([[${expression}]], ${request.id})`;
+        let p = request.promise.then(result => {
+            if (!result || result.length === 0) {
+                return;
+            }
+            let msg = result[0];
+            response.body = {
+                result: msg.result.toString(),
```
Comment (jschmidt42) [inline src/stingray-debugger.ts]: I am getting
```
Exception occured
TypeError: Cannot read property 'toString' of undefined
```

When I try to evaluate an local variable, i.e. `f5_pressed`

image

Comment (jschmidt42) [inline src/stingray-debugger.ts]: Is seems to be related to `isPotentialIdentifier` excluding identifiers with `_`.
Pushing a fix.

**Comment by jschmidt42** [comment]
Should this expression be expanded?
image

**Comment by jschmidt42** [comment]
@lochrist I have push the fix and re-ordered methods by scope/type.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,29 @@
 ## 1.0.0
 * First version
 
+## 1.1.0
+* Add better support for variable expansion.
+  ![image](https://cloud.githubusercontent.com/assets/4054655/24433504/30b184b4-13f7-11e7-98cd-e97c0eece92e.png)
+
+## TODO
```
Comment (jschmidt42) [inline CHANGELOG.md]: @lochrist FYI, move TODO items here

Comment (jschmidt42) [inline CHANGELOG.md]: See https://github.com/jschmidt42/stingray-vscode-debugger/blob/449d73e068b06181530a4cce296b1456afde78e6/CHANGELOG.md

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,29 @@
+## 1.1.0
+* Add better support for variable expansion.
+  ![image](https://cloud.githubusercontent.com/assets/4054655/24433504/30b184b4-13f7-11e7-98cd-e97c0eece92e.png)
+
 ## 1.0.0
 * First version
 
+## TODO
```
Comment (jschmidt42) [inline CHANGELOG.md]: @lochrist FYI, see https://github.com/jschmidt42/stingray-vscode-debugger/blob/525ac6b065c7210741b133e8b0f89ad1fbb82aa8/CHANGELOG.md

**Comment by jschmidt42** [comment]
If a table is empty, could we show it as an empty object?
image

image

**Pull request merged**